### PR TITLE
(WIP) => fix(date-picker): Updates avalible years to be dynamic to min and max

### LINKED ISSUE #10646: Hide years superior to max date in datepicker
#### Bug, feature request, or proposal:

Proposal, change the view years if a maxdate is set.

#### What is the expected behavior?

If i set the maxdate as today (in 2018) i don't expect to see 2019, 2020, 2021... in the years view.

#### What is the current behavior?

All the years are visible but 2019, 2020, 2021... are disabled.
issue_date

#### What is the use-case or motivation for changing an existing behavior?

I set a max date to limit the possibilities, i'm not (and the user) interrested in years that i can't select.
The extreme is if i set 2016, i can choose only this one among 24 differents years...

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?

Latest 

#### Is there anything else we should know?

I think the right disabled arrow if enough for understand that you can't go further in years.

### PR DESCRIPTION
**Pull request by raygervais**
This fixes #10646.

This removes the years which were originally greyed out. Brings in 4 edge cases:
1. No min / max: Standard Behavior
2. Only min: offset to 0 (display min year as first)
3. Only max: Display all paged years minus 1.
4. Both min and max: Same as 3.

### CODE REVIEW

**Comment by raygervais** [comment]
I will update with screens and progress tomorrow, just wanted to show initial thinking based on @WizardPC logic described in the bug and fix any syntax compliance issues.

**Comment by raygervais** [comment]
**Default behaviour (Pre Update where no min / max being provided):** 
screen shot 2018-04-20 at 11 03 10 am

**Providing a Min Date:** 
*Will fix heading which is displaying wrong year range*
screen shot 2018-04-20 at 11 04 22 am

I'm currently fixing the logic for case 3 and 4, which is broken. Opinions so far?

**Comment by mmalerba** [comment]
It seems reasonable to me, couple comments:
- For the first case, would it make more sense to put today's date somewhere near the middle? That way the user can go forward or back a couple years without changing pages. 
- The header in both images still says 2016-2039. I assume you just haven't gotten around to changing it yet?

**Comment by raygervais** [comment]
Working on your first point now, since I believe that will help lead to the correct design. I'm partially lost how to tackle case 3 & 4. And yes, I have not gotten around to changing the header yet. Any thoughts  / suggestions towards the header being change, or case 3 / 4?

**Comment by mmalerba** [comment]
I think the math would look something like this for the max year case:

If we were to put the current year at the end of the page, how many pages would we have to flip through to get to the max year? we can determine that like this: `currentYear + yearsPerPage * x = maxYear`

We want `x` to come out non-fractional here, because that means that the `maxYear` will also land at the end of the page. If `x` turns out fractional we can just move the `currentYear` back enough so that the `maxYear` lands at the end, like this: `currentYearOffsetFromEnd = (x - floor(x)) * yearsPerPage`.

If you want the offset from the beginning instead you can just do `currentYearOffsetFromStart = yearsPerPage - 1 - currentYearOffsetFromEnd` to get that.

**Comment by WizardPC** [comment]
@raygervais for the header, i think it's wil be the same but in `get periodButtonText()` from `/src/lib/datepicker/calendar.ts`.

I will try to find time to do a deeper analysis this week-end... be brave !

**Comment by raygervais** [comment]
Stupid question @mmalerba, perhaps it's been a long week but when you're referring to `x`, thats a value which we have to determine right? I need to find the best value be it 1, 200, etc.

**Comment by mmalerba** [comment]
Yeah `x` is something we need to calculate based on the max and the current year. `currentYear + yearsPerPage * x = maxYear`, so `x = (maxYear - currentYear) / yearsPerPage`

File: src/lib/datepicker/multi-year-view.ts [hunk 1]
```diff
@@ -132,7 +132,20 @@ export class MatMultiYearView<D> implements AfterContentInit {
...
     let activeYear = this._dateAdapter.getYear(this._activeDate);
+
+    // Default Behavior for Offset
     let activeOffset = activeYear % yearsPerPage;
+
+    if (!this._maxDate && this._minDate) {
+      activeOffset = 0;
+    }
+
+    if (!this._minDate && this._maxDate || this._minDate && this._maxDate) {
```
Comment (mmalerba) [inline src/lib/datepicker/multi-year-view.ts]: This can be simplified to `if (this._maxDate)`

File: src/lib/datepicker/multi-year-view.ts [hunk 2]
```diff
@@ -132,7 +132,20 @@ export class MatMultiYearView<D> implements AfterContentInit {
...
+      let yearOffset = activeYear - this._dateAdapter!.getYear(this._maxDate);
```
Comment (mmalerba) [inline src/lib/datepicker/multi-year-view.ts]: Prefer `const` over `let`. Also I don't think you need the `!` after `_dateAdapter`

File: src/lib/datepicker/multi-year-view.ts [hunk 3]
```diff
@@ -132,7 +132,20 @@ export class MatMultiYearView<D> implements AfterContentInit {
...
+      let currentYearOffsetFromEnd = (yearOffset - Math.floor(yearOffset)) * yearsPerPage;
+      activeOffset = (this._minDate) ? 0 : currentYearOffsetFromEnd;
```
Comment (mmalerba) [inline src/lib/datepicker/multi-year-view.ts]: Don't need `()` around `this._minDate`

**Comment by raygervais** [comment]
Still debugging, but one note that I noticed is that when your `_maxDate` is lesser than the current Date, then the `activeYear` is assigned the max's year. 
Ex: say `_maxDate` is `12/12/14`, then the `activeYear` is `2014`. This is what could be causing the offset to be sporadic at times.

I'm convinced this is partially the issue, looking more into. Unsure if that is a valid analysis / finding or if it doesn't relate at all.

**Comment by raygervais** [comment]
Here are the current use cases, which from my testing appear to be done (still need to correct / update header): 

1) No Min / Max Dates
screen shot 2018-04-22 at 4 01 24 pm

2) Only Min Date
screen shot 2018-04-22 at 4 01 43 pm

3) Only Max Date
screen shot 2018-04-22 at 4 02 33 pm

4) Both Min and Max Dates
screen shot 2018-04-22 at 4 02 12 pm

**Comment by raygervais** [comment]
Question, what is the best approach to handling the date-headings component? I'd like to complete this PR but at the moment I cannot think of a good approach for the last portion (updating the headings).

**Comment by mmalerba** [comment]
@raygervais sorry I haven't had a chance to look into this. I asked my intern to take a look, she created a new PR based off this plus a separate PR where I had fixed some change detection issues, feel free to comment on her PR here: #11485

**Comment by raygervais** [comment]
That's quite alright @mmalerba, honestly I was starting to come to terms that I had bitten off too much than I could handle at the time with quite a few out-side-the-computer events going on for me to truly dedicate focus and enthusiasm this direction. Happy to pass this on to your intern and continue contributing where possible in the close future!

I see that the Google Bot is saying that I have to authorize the use of my commits / branch being forked, how would I do so that there is no friction for anyone else working on it?

**Comment by mmalerba** [comment]
@raygervais if you could just comment on the PR saying that you're ok with your commits being used that would be great, thanks!

**Comment by JaegerCaiser** [comment]
I have a question about this, but fisrt just to be clear and understand the situation, what was propose is dont show the year  conform the min and max is setted?

Why dont just block the click in the year?

Because a have a situation that my selection date start in the year view, and I select the month at follow, and close datepicker to just select the year and month. But the click works same when the year is disabled.

**Comment by torabian** [comment]
You can use amazing time picker (amazing-time-picker) for this purpose, and it is fully compatible with Angular 2-7 and Material design:

https://github.com/owsolutions/amazing-time-picker

Live demo:
https://owsolutions.github.io/amazing-time-picker

Thanks.

**Comment by genyklemberg** [comment]
What is the status of this feature? When it can be used with proper headings?

**Comment by phl3x0r** [comment]
Status update?

**Pull request closed without merging**